src/supernova/mod.rs

```diff
@@ -0,0 +1,1232 @@
+//! This library implements SuperNova, a Non-Uniform IVC based on Nova.
+#![allow(dead_code)]
+
+use crate::{
+  constants::{BN_LIMB_WIDTH, BN_N_LIMBS, NUM_HASH_BITS},
+  errors::NovaError,
+  r1cs::{R1CSInstance, R1CSShape, R1CSWitness, RelaxedR1CSInstance, RelaxedR1CSWitness},
+  scalar_as_base,
+  traits::{
+    circuit::SuperNovaStepCircuit, circuit::SuperNovaTrivialTestCircuit,
+    commitment::CommitmentTrait, AbsorbInROTrait, Group, ROConstants, ROConstantsCircuit,
+    ROConstantsTrait, ROTrait,
+  },
+  Commitment, CommitmentKey,
+};
+
+use crate::gadgets::utils::alloc_num_equals;
+use crate::gadgets::utils::conditionally_select;
+use bellperson::gadgets::boolean::Boolean;
+use core::marker::PhantomData;
+use ff::Field;
+use ff::PrimeField;
+use serde::{Deserialize, Serialize};
+
+use crate::bellperson::{
+  r1cs::{NovaShape, NovaWitness},
+  shape_cs::ShapeCS,
+  solver::SatisfyingAssignment,
+};
+use ::bellperson::{Circuit, ConstraintSystem};
+use flate2::{write::ZlibEncoder, Compression};
+use sha3::{Digest, Sha3_256};
+
+use crate::nifs::NIFS;
+
+mod circuit; // declare the module first
+use circuit::{CircuitInputs, CircuitParams, SuperNovaCircuit};
+
+fn compute_digest<G: Group, T: Serialize>(o: &T) -> G::Scalar {
+  // obtain a vector of bytes representing public parameters
+  let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+  bincode::serialize_into(&mut encoder, o).unwrap();
+  let pp_bytes = encoder.finish().unwrap();
+
+  // convert pp_bytes into a short digest
+  let mut hasher = Sha3_256::new();
+  hasher.update(&pp_bytes);
+  let digest = hasher.finalize();
+
+  // truncate the digest to NUM_HASH_BITS bits
+  let bv = (0..NUM_HASH_BITS).map(|i| {
+    let (byte_pos, bit_pos) = (i / 8, i % 8);
+    let bit = (digest[byte_pos] >> bit_pos) & 1;
+    bit == 1
+  });
+
+  // turn the bit vector into a scalar
+  let mut digest = G::Scalar::ZERO;
+  let mut coeff = G::Scalar::ONE;
+  for bit in bv {
+    if bit {
+      digest += coeff;
+    }
+    coeff += coeff;
+  }
+  digest
+}
+
+/// A type that holds public parameters of Nova
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(bound = "")]
+pub struct PublicParams<G1, G2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+{
+  F_arity_primary: usize,
+  F_arity_secondary: usize,
+  ro_consts_primary: ROConstants<G1>,
+  ro_consts_circuit_primary: ROConstantsCircuit<G2>,
+  ck_primary: Option<CommitmentKey<G1>>,
+  r1cs_shape_primary: R1CSShape<G1>,
+  ro_consts_secondary: ROConstants<G2>,
+  ro_consts_circuit_secondary: ROConstantsCircuit<G1>,
+  ck_secondary: Option<CommitmentKey<G2>>,
+  r1cs_shape_secondary: R1CSShape<G2>,
```

Comment:
This seems to be precisely the same struct at `PublicParams`. Would some sizing of the (R1CS) contents of the ROM be more appropriate than `r1cs_shape_primary` and `r1cs_shape_secondary` here?

Reply:
`r1cs_shape_primary` and `r1cs_shape_secondary` are not the same struct, since secondary circuit is just a trivial circuit and do nothing while primary circuit got the application logic. Not sure I understand correctly, would you kindly elaborate more on this :)

Reply:
Setting heterogeneity aside for a moment, the sizing of the public parameters depends on the shape of the circuit they are used for. In particular, you would want the `CommitmentKey` (in practice, a vector Pedersen) to offer enough generators for the size of the step circuit. How this is done is expanded upon in PR #203 .

With heterogeneity, I would expect that same dependency to still exist, but the circuit bookkeeping to be different. More specifically, I would expect the `R1CSShape` of the *largest* circuit among those in the ROM to drive the size requirements for the public parameters.

Reply:
This was hinted at in a [prior comment](https://github.com/microsoft/Nova/pull/204#discussion_r1267328957). 
Nova, as a library, is quite safe in sizing public parameters:
- Nova runs through circuit synthesis (which is costly!) to size the public parameters precisely, see [here](https://github.com/microsoft/Nova/blob/f0b27b7457556f58cffb01e6c89b5c7e159a694f/src/lib.rs#L110-L121).
- see #203 for other nitty-gritty details.

The approach of simply recommending the user to pass the largest circuit among those in the ROM here may be error-prone -- as I'm worried a user might miss the requirement.

Reply:
> This was hinted at in a [prior comment](https://github.com/microsoft/Nova/pull/204#discussion_r1267328957). Nova, as a library, is quite safe in sizing public parameters:
> 
> * Nova runs through circuit synthesis (which is costly!) to size the public parameters precisely, see [here](https://github.com/microsoft/Nova/blob/f0b27b7457556f58cffb01e6c89b5c7e159a694f/src/lib.rs#L110-L121).
> * see [feat: Make commitment key generation configurable with an optional parameter #203](https://github.com/microsoft/Nova/pull/203) for other nitty-gritty details.
> 
> The approach of simply recommending the user to pass the largest circuit among those in the ROM here may be error-prone -- as I'm worried a user might miss the requirement. If there was a ROM API in the form of an indexed iterator of circuits (perhaps with a cached way to determine their shape), I suspect it would make automating this part of the setup possible.

For current version, the automating way to `identify largest primary circuit` is relying on user to implement logic to sizing the largest circuit to setup commitment key based on it. demo in [Supernova test](https://github.com/hero78119/SuperNova/blob/supernova/src/supernova/mod.rs#L1056-L1073) 
I admitted the ideal way is to exploring way to hide it from user to be more error pruning.